File: exam-frontend/src/services/auth.service.ts
import { apiClient } from '../config/api';
import { getBrowserFingerprint } from '../utils/fingerprint';
import { API_ENDPOINTS } from '../config/constants';
import type { UserRole } from '../types/auth';

export interface AuthResponseDto {
  accessToken: string;
  username: string;
  email: string;
  role: UserRole;
}

export const authService = {
  async login(username: string, password: string): Promise<AuthResponseDto> {
    const fingerprint = getBrowserFingerprint();
    const { data } = await apiClient.post<AuthResponseDto>(API_ENDPOINTS.AUTH.LOGIN, { username, password, fingerprint });
    return data;
  },

  async register(username: string, email: string, password: string): Promise<AuthResponseDto> {
    const fingerprint = getBrowserFingerprint();
    const { data } = await apiClient.post<AuthResponseDto>(API_ENDPOINTS.AUTH.REGISTER, { username, email, password, fingerprint });
    return data;
  },

  async logout(): Promise<void> {
    await apiClient.post(API_ENDPOINTS.AUTH.LOGOUT);
  }
};

File: exam-backend/src/main/java/com/exam/service/auth/AuthService.java
package com.exam.service.auth;

import com.exam.core.security.JwtProvider;
import com.exam.model.auth.RefreshSessionEntity;
import com.exam.model.auth.UserEntity;
import com.exam.model.enums.UserRole;
import com.exam.web.dto.auth.AuthResponseDto;
import com.github.benmanes.caffeine.cache.Cache;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {
    private final UserService userService;
    private final RefreshSessionService sessionService;
    private final JwtProvider jwtProvider;
    private final Cache<@NonNull String, Boolean> blacklistedTokensCache;

    @Transactional
    public AuthResult login(String username, String password, String fingerprint) {
        UserEntity user = userService.validateAndGetUser(username, password);
        RefreshSessionEntity session = sessionService.createSession(user, fingerprint);
        return new AuthResult(new AuthResponseDto(jwtProvider.generateAccessToken(user), user.getUsername(), user.getEmail(), user.getRole()),
                session.getRefreshToken());
    }

    @Transactional
    public AuthResult register(String username, String email, String password, String fingerprint) {
        UserEntity user = userService.createUser(username, email, password, UserRole.USER);
        RefreshSessionEntity session = sessionService.createSession(user, fingerprint);
        return new AuthResult(new AuthResponseDto(jwtProvider.generateAccessToken(user), user.getUsername(), user.getEmail(), user.getRole()),
                session.getRefreshToken());
    }

    @Transactional
    public AuthResult refresh(String oldToken, String fingerprint) {
        RefreshSessionEntity newSession = sessionService.rotateSession(oldToken, fingerprint);
        UserEntity user = newSession.getUser();
        return new AuthResult(new AuthResponseDto(jwtProvider.generateAccessToken(user), user.getUsername(), user.getEmail(), user.getRole()),
                newSession.getRefreshToken());
    }

    @Transactional
    public void logout(String accessToken, String refreshToken) {
        if (accessToken != null) {
            blacklistedTokensCache.put(accessToken, true);
        }
        sessionService.revokeSession(refreshToken);
    }

    public record AuthResult(AuthResponseDto responseDto, String refreshToken) {}
}
